### PR TITLE
fix(gomod): non-0.0.0 digest matching

### PR DESCRIPTION
**Pull request by omercnet**
Regex improved to match digest releases of tagged packages https://regex101.com/r/jaFaCr/1
https://github.com/renovatebot/renovate/discussions/19084

## Changes

When a go package on github is tagged with a version `vX.Y.Z` and a commit is pushed to its main branch without tagging, go mod detects the digest update as `vX.Y.Z+1-0.DATED_UPDATE-DIGEST`

This change allows detecting these types of updates

## Context
Fixes https://github.com/renovatebot/renovate/discussions/19084

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by omercnet** [comment]
tested in a private environment and seems to work

**Review by rarkins — CHANGES_REQUESTED** [review]
Please add some tests:
- They should *fail* if run against the current `main` branch
- They should pass if run against this PR

And did you have a public reproduction of the fix?

**Comment by omercnet** [comment]
hi @rarkins added a test and fixed an existing one that was failing

here's a public example
https://github.com/omercnet/renovate-go-test/issues/2 - official renovate app fails to detect a digest update
https://github.com/omercnet/renovate-go-test/issues/4 - my fixed version picks is up

**Comment by rarkins** [comment]
@omercnet is this what you expected? It changes from 0.0.2 to 0.0.0: https://github.com/omercnet/renovate-go-test/pull/19/files

**Comment by omercnet** [comment]
Yes - that's the most recent commit

**Comment by omercnet** [comment]
The important piece is detecting it as a legitimate digest package requirement

You can replicate this by doing `go get package@digest` and if it's untagged you'll get the same result

**Comment by rarkins** [comment]
Would there be people who might prefer that if the current digest is versioned (0.0.2) that they don't want to revert to 0.0.0 ?

**Comment by omercnet** [comment]
I don't fully understand the go kid versioning mechanism but that's the result of go get. 

Keep in mind this only happens when you decided to use a digest package. 

If you use v0.0.1 renovate (and go mod) won't offer an update

**Review by rarkins — APPROVED** [review]

File: lib/modules/datasource/go/index.ts [hunk 1]
```diff
@@ -59,8 +59,9 @@ export class GoDatasource extends Datasource {
       return null;
     }
 
-    // ignore v0.0.0- pseudo versions that are used Go Modules - look up default branch instead
-    const tag = value && !value.startsWith('v0.0.0-2') ? value : undefined;
+    // ignore vX.Y.Z-(0.)? pseudo versions that are used Go Modules - look up default branch instead
+    const tag =
+      value && !/^v\d+\.\d+\.\d+-0?\.?.*/.test(value) ? value : undefined;
```
Comment (viceice) [inline lib/modules/datasource/go/index.ts:64]: 😕 that regex is bad, it'll match any `1.2.3-...` like version

Comment (rarkins) [reply lib/modules/datasource/go/index.ts:64]: I guess our tests missed it, I will revert

Comment (omercnet) [reply lib/modules/datasource/go/index.ts:64]: why is this bad? if digest updates are not enabled it will not suggest updating it
the tests covered it

**Pull request merged**